src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -638,76 +655,95 @@ event::ptr primitive_inst::realloc_if_needed() {
-        updated_params.output_layouts[0] = updated_layout;
-    if (can_reuse_buffer) {
-        GPU_DEBUG_TRACE_DETAIL << id() << ": reuse previously allocated output buffer - "
-                               << actual_layout.get_buffer_size().count() << "/" << _max_output_layout_count
-                               << std::endl;
-        if (_outputs[0]->get_layout() != actual_layout) {
-            _outputs[0] = _network.get_engine().reinterpret_buffer(*_outputs[0], actual_layout);
+    for (size_t i = 0; i < actual_layouts.size(); ++i) {
+        bool can_reuse_buffer = (_outputs[i] && updated_layouts[i].get_buffer_size().count() <= _max_output_layout_count[i]);
+        std::pair<bool, ov::Shape> prealloc_info =
+            sp.predict_preallocation_shape(id(), updated_layouts[i], can_reuse_buffer, i, tmp_prealloc_count);
+        if (prealloc_info.first && sp.can_preallocate(ov::shape_size(prealloc_info.second) * (dt_sizes[i] / 8))) {
+            auto new_layout = updated_layouts[i];
+            new_layout.set_partial_shape(prealloc_info.second);
+            updated_params.output_layouts[i] = new_layout;
         }
-        if (need_reset_output_memory() && !can_be_optimized()) {
-            GPU_DEBUG_TRACE_DETAIL << id() << " : Need reset output memory considering user" << std::endl;
-            ev = _outputs[0]->fill(_network.get_stream());
+
+        if (updated_params.output_layouts[i].get_buffer_size().count() < updated_layouts[i].get_buffer_size().count()) {
+            updated_params.output_layouts[i] = updated_layouts[i];
         }
-        GPU_DEBUG_PROFILED_STAGE_MEMALLOC_INFO("reuse_buffer");
-    } else {
-        GPU_DEBUG_TRACE_DETAIL << id() << ": realloc output memory. "
-                               <<  " Current buffer_size=" << _max_output_layout_count
-                               <<  " Requested buffer_size=" << updated_layout.get_buffer_size().count() << std::endl;
-        _outputs = allocate_outputs(&updated_params, need_reset_output_memory(), true);
-        GPU_DEBUG_CODE(std::string memalloc_info = "");
-        GPU_DEBUG_CODE(for (size_t out_idx = 0; out_idx < _outputs.size(); ++out_idx) {
-            memalloc_info += (((_outputs.size() > 1) ? ("o" + to_string(out_idx) + ":") : "") +
-                              (_outputs[out_idx]->from_memory_pool ? "from_pool" : "new_alloc"));
-        })
-        GPU_DEBUG_PROFILED_STAGE_MEMALLOC_INFO(memalloc_info);
 
-        // TODO : need to handle multiple outputs
-        _max_output_layout_count = updated_params.output_layouts[0].get_buffer_size().count();
-    }
+        if (can_reuse_buffer) {
+            GPU_DEBUG_TRACE_DETAIL << id() << ": reuse previously allocated output buffer - "
+                                   << actual_layouts[i].get_buffer_size().count() << "/" << _max_output_layout_count[i]
+                                   << std::endl;
+            if (_outputs[i]->get_layout() != actual_layouts[i]) {
```

Comment:
I wonder if we can avoid this layouts comparison and just do reinterpret unconditionally? I think if may reduce overhead a little bit

Reply:
Removed.